src/flux/internal.c

```diff
@@ -0,0 +1,187 @@
+/***************************************************************************
+ *  Copyright (C) 2017, Lawrence Livermore National Security, LLC.
+ *  Produced at the Lawrence Livermore National Laboratory (cf, DISCLAIMER).
+ *  Written by Olaf Faaland <faaland1@llnl.gov>
+ *  UCRL-CODE-235649. All rights reserved.
+ *
+ *  This file is part of libyogrt.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this library.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ ***************************************************************************/
+
+
+#include <stdio.h>
+#include <jansson.h>
+#include <time.h>
+#include <limits.h>
+#include <flux/core.h>
+
+#include "internal_yogrt.h"
+
+#define BOGUS_TIME -1
+
+int jobid_valid = 0;
+int verbosity = 0;
+
+static flux_jobid_t jobid = 0;
+
+int internal_init(int verb)
+{
+	char *jobid_str;
+
+	verbosity = verb;
+	jobid_valid = 0;
+
+	if ((jobid_str = getenv("FLUX_JOB_ID")) == NULL) {
+		error("ERROR: FLUX_JOB_ID is not set."
+			  " Remaining time will be a bogus value.\n");
+		return jobid_valid;
+	}
```

Comment:
There is at least one important case I hadn't immediately thought of where a process is running in a Flux instance, but `FLUX_JOB_ID` is not set. When launched as a job under a foreign RM such as Slurm `FLUX_JOB_ID` will not be set, but the instance may have a timelimit from its resource manager job.

Other cases where `FLUX_JOB_ID` is not set for a Flux instance would be for the system instance and when running `flux start` standalone such as with `flux start -s, --test-size=N`. In both these cases it might make sense to return the equivalent of an "unlimited" timelimit.

Reply:
OK, good to know.
Is there a good way to know my process is running in a flux instance, in cases where there is no ```FLUX_JOB_ID``` environment variable set?  I was thinking it would be nice to tell the user "Hey, you got no Flux" in that case.  Perhaps call flux_open() and flux_attr_get() for some attribute?

Reply:
yeah, I think if `flux_open()` fails then it is fairly certain you are not running under Flux.

Reply:
Would below cover both the case where a job is running under a foreign RM and the case where there's no job, such as under flux start?

```
if (!(h = flux_open(NULL, 0))) {
    error("System not running flux");
    return -1;
}

if (!(exp = flux_attr_get(h, "expiration"))) {
    debug("Flux reports no time limit");
    return UNLIMITED_TIME
}

... parse the attribute received and calculate remaining seconds...
```

Reply:
I don't think there is currently a broker `expiration` attribute, though that is not a bad idea. The problem is that we'd need some kind of extensible system to discover the timelimit/expiration when Flux is launched under a foreign RM. I guess we could use libyogrt if available (:stuck_out_tongue_winking_eye:)

It would be beneficial to have a proper resource expiration for jobs launched under Slurm, etc. so that child jobs also had a correct time limit applied. I'll open an issue against flux-core if we don't already have one.

For now, could libyogrt attempt to fall back to another provider if we fail to get a timelimit from the current instance? If that fails, then perhaps an unlimited time could be returned (i.e. standalone `flux-start` case)

Reply:
Yes, when libyogrt can't get a time limit for any reason, it returns MAX_INT seconds to the yogrt_remaining() call.   So that part will be OK.

I'll have to look at the fallback idea.  Currently, libyogrt is told the provider at runtime by a config file (has to be built with support for the specified provider, of course), and the config file and the code allow for only one provider to be specified, as far as I know.  That could be changed, but I'm not sure how much work it would be and/or what side-affects there might be.  Seems less yucky than flux having to detect that it's running under slurm!